Completed Tests and CRUD

vramdhanie opened the pull request:
- CRUD functionality for all resources
- Authentication with JWT
- Tests

vramdhanie committed:
Merge pull request #2 from vramdhanie/feature/ci

Heroku deploy

vramdhanie committed:
Add Procfile to specify the web start for this porject.

vramdhanie committed:
Merge pull request #3 from vramdhanie/feature/ci

Add Procfile to specify the web start for this porject.

vramdhanie committed:
Revert "Add Procfile to specify the web start for this porject."

vramdhanie committed:
Merge pull request #4 from vramdhanie/revert-3-feature/ci

Revert "Add Procfile to specify the web start for this porject."

vramdhanie committed:
Fix broken styles to match Google style guide

vramdhanie committed:
Add custom exception handler for Note Exceptions

vramdhanie committed:
Add a Note Controller

vramdhanie committed:
Add Java configurations for exposing ids

vramdhanie committed:
Add delete endpoint for Notes

vramdhanie committed:
Add Tag controller

vramdhanie committed:
Add folder controller

vramdhanie committed:
Controllers for all entities

vramdhanie committed:
Change cascade type between note and tag. Add comments to explain purpose of extra methods

vramdhanie committed:
Add JsonIgnore to Tag side of Note <--> Tag association to avoid infinite loop in JSON converter

vramdhanie committed:
Note Controller asociations to tags and folders

vramdhanie committed:
Add Note controller tests

vramdhanie committed:
Add createNote test

vramdhanie committed:
Add jwt dependency

vramdhanie committed:
Add User controller and Repository for creating new users

vramdhanie committed:
Add BCrypt password encoder bean

vramdhanie committed:
Add UserController

vramdhanie committed:
Implement the Authentication itself.

Two filters: JwtAuthentication and JwtAuthorization act like Basic and JWT Strategy in passport.

vramdhanie committed:
Merge pull request #5 from vramdhanie/feature/authentication

Feature/authentication

vramdhanie committed:
Also deploy from preview branch

vramdhanie committed:
Add REST doc dependencies for future use

vramdhanie committed:
Update CRUD operations to consider the logged in User

vramdhanie committed:
Update tests to use Mock User for authentication

vramdhanie committed:
Add API advice to README

vramdhanie committed:
Merge pull request #6 from vramdhanie/feature/tests

Feature/tests

benjaminEwhite commented:
Looks java-licious to me :)

@Thoughtscript could I ask you to review this PR?

Thoughtscript commented on src/main/java/com/thinkful/noteful/auth/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,71 @@
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.springframework.beans.factory.annotation.Autowired;
```
`import org.springframework.beans.factory.annotation.Autowired;` is no longer needed - smoke it.

Thoughtscript commented on src/main/java/com/thinkful/noteful/auth/JwtAuthorizationFilter.java:
```diff
@@ -0,0 +1,63 @@
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.springframework.beans.factory.annotation.Autowired;
```
`import org.springframework.beans.factory.annotation.Autowired;` - looks like it can be dropped.

vramdhanie replied:
Done

Thoughtscript commented on src/main/java/com/thinkful/noteful/auth/WebSecurityConfig.java:
```diff
@@ -0,0 +1,65 @@
+
+import static com.thinkful.noteful.auth.SecurityConstants.SIGN_UP_URL;
+
+import com.thinkful.noteful.users.UserRepository;
```
`import com.thinkful.noteful.users.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;` - we can remove both of these too!

vramdhanie replied:
Done

Thoughtscript commented on src/main/java/com/thinkful/noteful/tags/TagDeserializer.java:
```diff
@@ -0,0 +1,36 @@
+
+  @Override
+  public Tag deserialize(JsonParser jp, DeserializationContext ctx) 
+        throws IOException, JsonProcessingException {
```
`JsonProcessingException` is a subclass of `IOException` so we can drop it without weakening our error-handling here.

vramdhanie replied:
@Thoughtscript This is a good point. However, the deserialize() method in the super class declares it this way so I usually try to maintain the structure provided by the super class for clarity, even though listing JsonProcessingException does not add anything to the code. What are your thoughts on that? I'll be happy to remove it.

Thoughtscript replied:
@vramdhanie - Sure, it's unneeded from a programmatic standpoint but does help to illustrate using the newer multi-catch syntax.

Thoughtscript commented on src/main/java/com/thinkful/noteful/auth/WebSecurityConfig.java:
```diff
@@ -0,0 +1,65 @@
+      .csrf()
+      .disable()
+      .authorizeRequests()
+      .antMatchers(HttpMethod.POST, SIGN_UP_URL).permitAll()
```
Wasn't sure if this was completed or if I'm missing something but wanted to verify that the SIGN_UP_URL permits user creation and sign up?

![sign up v2](https://user-images.githubusercontent.com/10905039/39965048-8abe45f4-5646-11e8-8657-8d10574c7e4c.PNG)

vramdhanie replied:
Yes, this is done, you should be able to create a new user with the sign up url. I have examples in the README.md file.

Thoughtscript commented on src/main/java/com/thinkful/noteful/folders/Folder.java:
```diff
@@ -16,81 +21,87 @@
-    }
+  public void setUser(User user) {
+    this.user = user;
+  }
```
Hibernate and Domain side look great!

![capture](https://user-images.githubusercontent.com/10905039/39965054-b0552bca-5646-11e8-87c9-5283a4357d82.PNG)

Thoughtscript commented on src/main/java/com/thinkful/noteful/folders/FolderController.java:
```diff
@@ -0,0 +1,92 @@
+    folderRepository.delete(folder);
+    return folder;      
+  }
+
```
Controllers look great (though I haven't been able to authenticate yet):

![capture2](https://user-images.githubusercontent.com/10905039/39965061-cd0cdbbe-5646-11e8-802d-bba975c274e3.PNG)

vramdhanie replied:
Authentication is possible by first POST /api/users/sign-up with 

```
{
    "username":"fred",
    "password":"flintstone"
}
```
then login with POST /login and same credentials. I updated the login to return the authToken in the body as expected by the sample client application as well as in a header.

Thoughtscript commented on README.md:
```diff
@@ -35,16 +35,80 @@ gradle bootRun
 This starts the embedded Tomcat server on port 8080.
 
 
-~~**NOTE**: A single user with username 'samwise' and password 'gamgee' has been configured in the _application.properties_ file.~~ **Security temporarily disabled**
```
Adding: `spring.jpa.properties.hibernate.temp.use_jdbc_metadata_defaults=false` into `application.properties` will prevent the following from appearing at startup (as you know, it's an ignorable error but it might help to avoid some confusion):

![exception](https://user-images.githubusercontent.com/10905039/39965098-cbfa4fe4-5647-11e8-8f23-cabd26947495.png)

Thoughtscript reviewed (changes requested):
Overall great work - security / authentication could use some clarification on endpoints and process when running locally. Minor tweak to `application.properties` recommended. Cheers!

vramdhanie committed:
Remove extraneous imports

vramdhanie committed:
remove extraneous imports

vramdhanie committed:
Add hibernate config to disable error message on startup

vramdhanie committed:
remove extraneous imports

vramdhanie committed:
Set authorization headers for CORS

vramdhanie committed:
reset configurations

vramdhanie committed:
Modified Authentication to return the authToken in the body

vramdhanie committed:
Set deploy from branch staging only

vramdhanie committed:
Add String constructors for Tags and Folders as a convenience for HTTP transport

vramdhanie committed:
Change login URL to POST /api/login

vramdhanie committed:
Updated CORS config to allow pre - flight requests

vramdhanie committed:
Changed user sign-up url to 'POST /api/users'

vramdhanie committed:
Add exception handling that matches API specs better.

Validation exceptions are now returned with the same JSON structure os the spec.

vramdhanie committed:
Should return 401 instead for 403 for authentication errors and avoid duplicate usernames

vramdhanie committed:
Add validation for all controllers

vramdhanie committed:
Add custom serializer for Note to match API spec

vramdhanie committed:
Add custom serializers for Tags and folders

vramdhanie committed:
Disable tests temporarily for deployment

Thoughtscript commented:
Looks like some great improvements all around - I've tried to organize my comments into the three areas to enhance our velocity here. These are listed in no particular order and given that items are currently in dev, the order in which they are completed is of course whatever works best.

OK, from a testing standpoint these look like good scaffolds and starting points - we'll want to make sure that:

1.	`NoteControllerTest` could be expanded to provide full coverage over the controllers. Ditto for each of the other controllers!
2.	`NotefulApplicationTests` is currently looking great from a boilerplate standpoint. Excited to see this incorporate a full integration test!
3.	Would be great to see some of authentication controller tests.

I have several boilerplate tests (integration and unit) here if you find them suitable: https://github.com/Thoughtscript/java_stuff

**Authentication:**

![properties](https://user-images.githubusercontent.com/10905039/40630415-63e525f4-6287-11e8-8c71-2fce47ab43a6.PNG)

Per: https://docs.spring.io/spring-boot/docs/current/reference/html/common-application-properties.html and `application.yml`:

```
spring.security.user.name = samwise
spring.security.user.password = gamgee
```

I should be able to authenticate with that default user.

![ezgif-5-1a62b771d0](https://user-images.githubusercontent.com/10905039/40630422-72eaf150-6287-11e8-9317-3cf385ba5211.gif)

4. I am not able to sign up with a new user - u: frodo, p: baggins on heroku either using SSL or no SSL
5. I am not able to sign up with a new user - u: frodo, p: baggins locally either using SSL or no SSL
6. I am not able to login in with the default user - u: samwise, p: gamgee on heroku either using SSL or no SSL
7. I am not able to login in with the default user - u: samwise, p: gamgee locally either using SSL or no SSL

![localhost](https://user-images.githubusercontent.com/10905039/40630420-6cc92bde-6287-11e8-8ab8-d66b940af4e5.PNG)

![postgres](https://user-images.githubusercontent.com/10905039/40630425-79e8db52-6287-11e8-945c-0377dcaa4eee.PNG)

**Error Handling:**

It'd be nice to enrich our methods and logging to provide greater detail in light of these listed points.

8. Great work improving the Response data in each controller. However, when I try to authenticate either through login or signup or try to access a protected resource I get now response information.
9. Logging in strong but I also see no information regarding failed authentication attempts - it remains opaque to me (perhaps I'm doing something wrong) - my Postgres environment also remains empty.

Please let me know if I misunderstood or missed anything - overall it's looking good - the comments above indicate some of the last items before this is production-worthy. Cheers!

vramdhanie commented:
@Thoughtscript Thanks for this detailed response. I repeated what you tried and I see that you are right about Postman. For some reason the authentication is not working with postman. However, it is working from the browser using the [client app](https://github.com/Thinkful-Ed/noteful-app-client.git). Using the client app all the functionality works with both the deployed API on https://damp-coast-35809.herokuapp.com and running the server locally on http://localhost:8080. 

I'm not sure why this behaviour so I'll look into it. I'll also add some further tests as suggested.

vramdhanie commented:
@Thoughtscript I found the problem, but I'm not sure of the solution as yet. To create a new user ensure that the URL ends with a slash like this: http://localhost:8080/api/users/. Without that trailing slash it goes bananas. 

Now that happens because of the SIGN_UP_URL which is used to expose this particular URL while protecting everything else. However, if I try to remove the slash the default POST /api/users provided by Spring Rest will kick in and the password will not be encrypted and the JWT filter will break. I am looking to see how I can disable the default /api/users but in the mean time just use the trailing slash to test the functionality.

vramdhanie committed:
Allow create user urls to end with and without slashes, removed unneeded default CRUD operations on User

vramdhanie commented:
@Thoughtscript Alright, I fixed it. You should be able to get through to POST /api/users and POST /api/users/ now. twas a bit of a workup.

Let me know if this solves the problem for you.

Thoughtscript commented:
Hey @vramdhanie - I'll take a look here shortly! Thank you!

Thoughtscript commented:
OK, sorry for the early morning comments. Looks like lots of progress. Thank you for fleshing out the endpoints.

Two questions:

1. Still having some issues getting back a valid token either on the Heroku app or locally:

![login and signup](https://user-images.githubusercontent.com/10905039/40885170-16c53bbc-66d6-11e8-9103-9e5b7700da6e.png)

(**Heroku**)

![local login and signup](https://user-images.githubusercontent.com/10905039/40885177-2214b1b4-66d6-11e8-8d34-8c8756adc3e6.png)

(**Locally**)

Much improved just not generating the Authorization token. Those were tested with CURL and Postman - same results in either case. Local instance must be tested without SSL.

2. Logistically, did you envision multiple PR's? Or just this one. I'm happy merging this one and having one more to polish up. 

Cheers!

vramdhanie commented:
Hey @Thoughtscript Thanks for the hard work on this. I had changed the sign-up URL to match the URL used in the spec implementation. That is, POST /api/users . I think that may have caused some confusion. I'll double check my docs and make sure I update it wherever it may be wrong. 

I think we can finish the review on this PR then just merge it when done.

Thoughtscript commented:
@vramdhanie - Great, I've appreciated working with you on this. It's at a better place and I look forward to reviewing any additional polishing touches.
